netwayste/src/utils/ping.rs

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (C) 2020 The Conwayste Developers
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation, either version 3 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of  MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+use std::collections::VecDeque;
+use std::time::Instant;
+
+const PING_FILTER_DEPTH: usize = 12;
+
+pub struct PingFilter {
+    pub average_latency_ms: u128,
+    running_sum: u128,
+    history: VecDeque<u128>,
+    start_timestamp: Instant,
+    in_progress: bool,
+}
+
+impl PingFilter {
+    pub fn new() -> PingFilter {
+        PingFilter {
+            average_latency_ms: 0,
+            running_sum: 0,
+            history: VecDeque::with_capacity(PING_FILTER_DEPTH),
+            start_timestamp: Instant::now(),
+            in_progress: false,
+        }
+    }
+
+    pub fn reset(&mut self) {
+        let Self {
+            ref mut average_latency_ms,
+            ref mut running_sum,
+            ref mut history,
+            ref mut start_timestamp,
+            ref mut in_progress,
+        } = *self;
+
+        *average_latency_ms = 0;
+        *running_sum = 0;
+        history.clear();
+        *start_timestamp = Instant::now();
+        *in_progress = false;
+    }
+
+    pub fn start(&mut self) {
+        self.start_timestamp = Instant::now();
+        self.in_progress = true;
+    }
+
+    pub fn update(&mut self) {
+        if !self.in_progress {
+            panic!("start() was not called so a duration cannot be computed");
+        }
+
+        let latency = Instant::now().duration_since(self.start_timestamp);
+        let latency_ms = latency.as_millis();
+
+        self.running_sum += latency_ms;
+        self.history.push_back(latency_ms);
+
+        // Wait for the filter to be populated
+        if self.history.len() > PING_FILTER_DEPTH {
+            // unwraps safe b/c of length check
+            let oldest = self.history.pop_front().unwrap();
+            self.running_sum -= oldest;
+            self.average_latency_ms = (self.running_sum as f64 / PING_FILTER_DEPTH as f64) as u128;
```

Comment:
If you used `self.history.len()` as the denominator you could do this on every `update()` call rather than only when the history is full. :sunglasses:

Reply:
See common below regarding a meaningful number of samples. It's not so much as being "full", but instead as having "enough".